File: frontend/src/services/api/pipeline.ts
// Orchestration Cockpit API client (F-007, CR-NS-018 Phase 4).
//
// Backend owns the pipeline state; the cockpit is a thin board over it. Types
// mirror backend/schemas/pipeline.py exactly.

import api from "../api";
import type { components } from "./pipeline.generated";
import type { PaginatedResponse } from "../../types/common";

// Debug-attach (CR-V2-015) break-glass targets a v2 orchestrator session — the two agents only. These are
// CHARTER-PATH SLUGS (hyphen, e.g. `ai-agent`), which the BE bridges to the DB role value (underscore,
// `ai_agent`) via `db_role_for_charter_slug`; do not pass the underscore form to the debug-terminal route.
export type DebugAttachRole = "ai-agent" | "auditor";

// ── stage / actor / status enums (GENERATED — single source of truth, v0.7.0 R2) ─────────────
//
// These unions are no longer hand-mirrored from the backend CHECK constraints; they are derived
// from `pipeline.generated.ts` (produced by `npm run codegen` ← OpenAPI ← the Pydantic `Literal`
// schemas in backend/schemas/pipeline.py). `pipeline.ts` stays the stable import surface, so
// consumers (the cockpit panels et al.) import these names unchanged while the VALUES track the
// backend automatically — a BE change that isn't regenerated fails the CI drift-gate (R2-c).
type PipelineSchemas = components["schemas"];

export type PipelineStage = PipelineSchemas["PipelineStateRead"]["current_stage"];

export type PipelineActor = PipelineSchemas["PipelineStateRead"]["current_actor"];

// `manazer` (the human operator) and `system` are message-only participants (never a state ACTOR — the
// actor union is the two agents). The BE keeps author/recipient as plain str, so this composite stays
// hand-written over the generated actor union + the two human/system authors.
export type PipelineParticipant = PipelineActor | "manazer" | "system";

export type PipelineStatus = PipelineSchemas["PipelineStateRead"]["status"];

// R4 (D1): the persisted reason a pipeline is `blocked` (generated → single source with the BE Literal +
// DB CHECK). `NonNullable` strips the `| null` so the union is the four canonical values.
export type BlockReason = NonNullable<PipelineSchemas["PipelineStateRead"]["block_reason"]>;

export type PipelineMessageKind = PipelineSchemas["PipelineMessageRead"]["kind"];

// ── row types ────────────────────────────────────────────────────────────────

export interface PipelineState {
  id: string;
  version_id: string;
  flow_type: string;
  current_stage: PipelineStage;
  current_actor: PipelineActor;
  status: PipelineStatus;
  next_action: string;
  is_regate: boolean;
  iteration: number;
  // R4 (D1): why the pipeline is `blocked` — authoritative; the banner + action bar derive question-vs-error
  // from this. `null`/absent (legacy rows, or not blocked) → the FE falls back to the `isErrorBlock` heuristic.
  block_reason?: BlockReason | null;
  created_at: string;
  updated_at: string;
}

// Per-agent liveness for the who's-up staleness chips (CR-V2-021 — v2: the two agents only).
export type AgentLiveness = "idle" | "active" | "stale";
export interface AgentSession {
  role: PipelineActor;
  status: AgentLiveness;
}

export interface PipelineMessage {
  id: string;
  version_id: string;
  stage: PipelineStage;
  author: PipelineParticipant;
  recipient: PipelineParticipant;
  kind: PipelineMessageKind;
  content: string;
  status: string;
  payload: Record<string, unknown> | null;
  created_at: string;
  // Monotonic insertion order (CR-NS-018) — authoritative ordering for both the REST
  // board and incremental WS message_added frames; sort by it, don't trust arrival.
  seq: number;
}

export interface PipelineBoard {
  state: PipelineState | null;
  recent_messages: PipelineMessage[];
  // Backend-authoritative set of schvaľovacie-body actions valid to offer right now (WS-C1, CR-NS-030;
  // dial-governed v2 verbs). The action bar renders only these; absent → fall back to the FE's own logic.
  available_actions?: PipelineActionName[];
  // Build-readiness facts (WS-C1, CR-NS-030): the FE disables the Programovanie sign-off button when not
  // satisfiable (all_tasks_done false → a todo remains; build_open_findings > 0 → a failed/unverified task)
  // and drives the Programovanie split-view task progress. Absent → permissive (don't disable).
  all_tasks_done?: boolean;
  build_open_findings?: number;
  // The Programovanie task currently in focus (WS-C2, CR-NS-035) — the who's-up status shows "#N: title".
  current_task?: { number: number; title: string } | null;
  // Per-agent liveness for the who's-up staleness chips (the two v2 agents). Absent → no indicator.
  agent_sessions?: AgentSession[];
  // CR-V2-056 (reality-anchoring): is this version VERIFIED right now, COMPUTED live from the repo (the PASS
  // verdict's bound commit SHA vs the current HEAD), not a stored 'done'. `verified_provenance === 'sha_drift'`
  // means it WAS verified but the code moved past the verified commit — the board flags a stale-PASS warning
  // instead of a frozen green. Absent (older board) → no drift indicator.
  verified?: boolean;
  verified_provenance?: string;
}

// ── action requests ──────────────────────────────────────────────────────────

// The dial-governed v2 schvaľovacie-body verbs (CR-V2-009; mirrors orchestrator._ACTIONS). The v1
// Gate-E / gate_g / Coordinator / release action verbs (approve / return / fix / leave / verdict-PASS-gate
// / apply_coordinator_recommendation / end_gate_e / end_build / continue_build / rerun_release_audit /
// surgical_fix / uat_accept / retry_publish / accept_merged) are dropped with the 4-phase model.
export type PipelineActionName =
  | "start" // Spustiť tvorbu špecifikácie
  | "approve_spec" // Schváliť špecifikáciu (end Príprava — ALWAYS mandatory, dial-independent)
  | "schvalit" // Schváliť (dial-governed advance after Návrh / Programovanie / Verifikácia)
  | "uprav" // Uprav (rework the current phase / "Skús znova" on an error block)
  | "pokracovat" // Pokračovať (resume a paused Programovanie loop)
  | "verdict" // the Auditor's Verifikácia verdict (PASS / FAIL)
  | "ask" // open a direct AI-Agent consult
  | "answer" // answer an agent QUESTION on a blocked state
  | "pause" // cooperative pause of the Programovanie loop
  | "decide" // CR-V2-041: pick one consultation Decision Card option (decision_needed)
  | "overit_znovu"; // CR-V2-057: "Over znova" — re-verify a drifted version (re-run Verifikácia vs current HEAD)

export interface PipelineActionRequest {
  action: PipelineActionName;
  payload?: Record<string, unknown>;
}

// Minimal shape of the agent-terminal row returned by the debug-attach endpoint.
export interface DebugTerminalSession {
  id: string;
  role: DebugAttachRole;
  project_slug: string;
  pid: number;
}

// Fast-Fix Lane (F-009, CR-NS-094/095): the result of the "Rýchla oprava" one-prompt entry — the
// auto-created PATCH version's id + its initial board snapshot. Mirrors backend FastFixStartResponse.
export interface FastFixStartResponse {
  version_id: string;
  board: PipelineBoard;
}

// CR-V2-015 / SPIKE-IO Model B: the result of relaying a Manažér message to the AI Agent. The engine is the
// SOLE writer to the warm `claude` session — the message becomes the next `--resume` turn (never a raw
// keystroke). `deferred === true` ⇒ a turn was in flight and the message was ENQUEUED behind it (it lands at
// the next turn boundary); `false` ⇒ it dispatched immediately. Mirrors backend PipelineRelayResponse.
export interface PipelineRelayResponse {
  deferred: boolean;
  board: PipelineBoard;
}

// ── REST ─────────────────────────────────────────────────────────────────────

export function getPipelineBoardApi(versionId: string, limit = 50): Promise<PipelineBoard> {
  return api.get<PipelineBoard>(`/pipeline/${versionId}?limit=${limit}`);
}

// Fast-Fix Lane entry (F-009 §3, CR-NS-095): one prompt → the backend auto-creates the next PATCH
// version (vX.Y.Z+1) and starts a `fast_fix` pipeline carrying the Director directive. Returns the new
// version_id (navigate the cockpit to it) + the initial board.
export function startFastFixApi(projectId: string, directive: string): Promise<FastFixStartResponse> {
  return api.post<FastFixStartResponse>("/pipeline/fast-fix", {
    project_id: projectId,
    directive,
  });
}

export function listPipelineMessagesApi(
  versionId: string,
  skip = 0,
  limit = 50,
): Promise<PaginatedResponse<PipelineMessage>> {
  return api.get<PaginatedResponse<PipelineMessage>>(
    `/pipeline/${versionId}/messages?skip=${skip}&limit=${limit}`,
  );
}

export function postPipelineActionApi(
  versionId: string,
  body: PipelineActionRequest,
): Promise<PipelineBoard> {
  return api.post<PipelineBoard>(`/pipeline/${versionId}/action`, body);
}

export function openDebugTerminalApi(
  versionId: string,
  role: DebugAttachRole,
): Promise<DebugTerminalSession> {
  return api.post<DebugTerminalSession>(
    `/pipeline/${versionId}/debug-terminal?role=${encodeURIComponent(role)}`,
  );
}

// CR-V2-015 / SPIKE-IO Model B: relay a Manažér message to the AI Agent as the engine's next turn. This is
// the canonical Manažér→AI-Agent channel for the read-only AI Agent tab — the message is RELAYED by the
// engine (the sole writer to the warm `claude` session), never keystroked into the PTY. `deferred === true`
// means a turn was in flight and the message is queued behind it.
export function relayPipelineMessageApi(
  versionId: string,
  text: string,
): Promise<PipelineRelayResponse> {
  return api.post<PipelineRelayResponse>(`/pipeline/${versionId}/relay`, { text });
}

// ── WebSocket ────────────────────────────────────────────────────────────────

export function buildPipelineWsUrl(versionId: string, token: string): string {
  const protocol = window.location.protocol === "https:" ? "wss:" : "ws:";
  const base =
    (import.meta.env.VITE_API_BASE_URL as string | undefined) ||
    `${protocol}//${window.location.host}`;
  const wsBase = base.replace(/^http/, "ws");
  return `${wsBase}/api/v1/pipeline/ws/${versionId}?token=${encodeURIComponent(token)}`;
}

// One line of live agent activity (ephemeral — never persisted).
export interface ActivityLine {
  stage: PipelineStage;
  actor: PipelineActor;
  // "status" = a per-turn active-role signal ("…pracuje") that steps the rail
  // through the agents in a round (CR-NS-018).
  kind: "tool" | "text" | "status" | "";
  line: string;
}

// CR-V2-018 Helpers feed — the AI Agent's ephemeral sub-agent (Task) spawns, captured from the stream-json
// and broadcast over the pipeline WS whenever the active set changes. `count === 0` ⇒ the panel HIDES (the
// last helper finished). `line` is the Slovak "+ N pomocníci" header; `helpers` are the per-helper one-liners
// (spawn order). The Auditor is never a helper (independence — enforced backend-side).
export interface HelpersFeed {
  stage: PipelineStage;
  count: number;
  line: string;
  helpers: string[];
}

// WS frame shapes pushed by the backend (backend/api/routes/pipeline.py + services/pipeline_runner.py).
export type PipelineWsFrame =
  | { type: "state_changed"; board: PipelineBoard } // initial snapshot on connect
  | { type: "state_changed"; state: PipelineState } // delta after an action
  | { type: "message_added"; message: PipelineMessage }
  | ({ type: "agent_activity" } & ActivityLine) // live stream while agent_working
  | ({ type: "helpers" } & HelpersFeed) // CR-V2-018: ephemeral helper feed (panel hides at count 0)
  // CR-V2-015: the raw-PTY single-writer guard frame. The AI Agent tab relays through the engine (never raw
  // keystrokes), so this is mainly the break-glass console's signal; surfaced here so a frame on the shared
  // socket is handled gracefully rather than silently dropped.
  | { type: "write_rejected"; reason: string };

File: backend/schemas/pipeline.py
"""Pydantic schemas for the pipeline cockpit API (F-007 §6, CR-NS-018 Phase 3)."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from backend.db.models.pipeline import (
    ACTOR_VALUES,
    BLOCK_REASON_VALUES,
    FLOW_TYPE_VALUES,
    MESSAGE_KIND_VALUES,
    STAGE_VALUES,
    STATUS_VALUES,
)

# Literal aliases sourced from the DB CHECK value tuples (v0.7.0 R2, D2) — the response-schema enums
# and the DB constraints share ONE source so they cannot drift. FastAPI introspects each ``Literal``
# into an OpenAPI ``enum``, which drives the generated FE pipeline types (R2-b). ``Literal[<tuple>]``
# is equivalent at runtime to spelling the members out (``Literal["a", "b"]`` already passes a tuple),
# and Pydantic validates them; the backend uses no static type-checker (intentional), so the dynamic
# construction needs no ``# type: ignore``.
FlowType = Literal[FLOW_TYPE_VALUES]
PipelineStage = Literal[STAGE_VALUES]
PipelineActor = Literal[ACTOR_VALUES]
PipelineStatus = Literal[STATUS_VALUES]
BlockReason = Literal[BLOCK_REASON_VALUES]
MessageKind = Literal[MESSAGE_KIND_VALUES]


class PipelineStateRead(BaseModel):
    """Serialised ``pipeline_state`` row — "who is on turn and what's next"."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    version_id: UUID
    flow_type: FlowType
    current_stage: PipelineStage
    current_actor: PipelineActor
    status: PipelineStatus
    next_action: str
    is_regate: bool
    iteration: int
    #: R4 (D1): why the pipeline is ``blocked`` (authoritative; replaces the FE ``isErrorBlock`` heuristic).
    #: ``None`` whenever ``status != 'blocked'`` (and on legacy blocked rows pre-067 → FE heuristic fallback).
    block_reason: Optional[BlockReason] = None
    created_at: datetime
    updated_at: datetime


class PipelineMessageRead(BaseModel):
    """Serialised ``pipeline_message`` row (append-only log entry)."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    version_id: UUID
    stage: str
    author: str
    recipient: str
    kind: MessageKind
    content: str
    status: str
    #: Free-form per-message structured data (deliverables / commits / findings / usage / timing / the
    #: synthesis & autonomous markers …). Legible-cockpit-output fix adds ``report``: the agent's full
    #: human-readable markdown report body — the rich source the cockpit bubble renders, while ``content``
    #: stays the one-line summary. Additive within the existing freeform dict — no schema / codegen change.
    payload: Optional[dict[str, Any]] = None
    created_at: datetime
    #: Monotonic insertion order (CR-NS-018). Carried in the payload so both the REST
    #: board and the incremental WS broadcast expose the authoritative order — clients
    #: can sort by it instead of relying on arrival timing.
    seq: int


class BoardTask(BaseModel):
    """The Programovanie task currently in focus, for the who's-up status (WS-C2, CR-NS-035; v2 §4.4.2)."""

    number: int
    title: str


class AgentSession(BaseModel):
    """Per-agent liveness for the who's-up status — ``idle`` / ``active`` / ``stale`` from R1's
    ``OrchestratorSession.last_input_at`` heartbeat. v2: only the two agents (AI Agent / Auditor)."""

    role: str
    status: Literal["idle", "active", "stale"]


class PipelineBoardRead(BaseModel):
    """Vývoj board snapshot: current 4-phase state + the most recent messages (CR-V2-021).

    The v2 Vývoj board (design §4.4.2) renders a horizontal 4-phase bar (Príprava → Návrh → Programovanie
    → Verifikácia → Hotovo) whose chips ARE the tabs; each phase's durable artifact (Špecifikácia .md /
    design doc / coding log / Auditor verdict) is read by the FE from ``recent_messages`` (the phase
    gate_report / verdict carries it in ``payload['report']``). ``state`` is ``None`` until ``start``.

    The v1 Gate-E / gate_g / Coordinator board fields (``gate_e_open_findings`` / ``release_acceptance_
    satisfied`` / ``regate_proposal`` / ``coordinator_triage`` / ``autonomous_decisions_summary``) are
    DROPPED — there is no Gate E, no gate_g release-gate, and no Coordinator hub in the 4-phase model.
    """

    state: Optional[PipelineStateRead] = None
    recent_messages: list[PipelineMessageRead] = Field(default_factory=list)
    #: Backend-authoritative set of schvaľovacie-body actions valid to OFFER right now (WS-C1, CR-NS-030;
    #: rebuilt to the dial-governed v2 verbs in CR-V2-009 — ``approve_spec`` / ``schvalit`` / ``uprav`` /
    #: ``pokracovat`` / ``ask`` / ``answer`` / ``verdict`` / ``pause``). The FE renders only these; empty when
    #: the pipeline hasn't started.
    available_actions: list[str] = Field(default_factory=list)
    #: Build-readiness facts (WS-C1, CR-NS-030) the FE uses to DISABLE the Programovanie sign-off button when
    #: not satisfiable: ``all_tasks_done`` False → a task is still ``todo``; ``build_open_findings`` > 0 → a
    #: failed/unverified task. Defaults are the permissive "ready" values so an absent field never disables.
    #: Also feeds the Programovanie split-view task progress (design §4.5).
    all_tasks_done: bool = True
    build_open_findings: int = 0
    #: The Programovanie task currently in focus (WS-C2, CR-NS-035) — in_progress while coding, else the held
    #: failed task at a fix-loop HALT; the who's-up status shows "#N: title". ``None`` outside Programovanie.
    current_task: Optional[BoardTask] = None
    #: Per-agent liveness (idle / active / stale) for the who's-up status — one entry per v2 agent role
    #: (AI Agent / Auditor). An absent field on an older board → the FE renders no staleness indicator.
    agent_sessions: list[AgentSession] = Field(default_factory=list)
    #: CR-V2-056 (reality-anchoring): is this version VERIFIED *right now*, COMPUTED from the live repo (the
    #: Verifikácia PASS verdict's bound commit SHA vs the current HEAD) — NOT a stored 'done' snapshot. Lets
    #: the board reflect reality: a version whose HEAD drifted past its verified commit shows a stale-PASS
    #: warning instead of a frozen green. ``verified_provenance`` is the reason: ``sha_match`` (verified) /
    #: ``sha_drift`` (was verified, HEAD moved past it — the frozen PASS the FE flags) / ``unbound`` /
    #: ``legacy`` / ``repo_unreadable`` / ``no_pass`` (never passed / re-judge pending).
    verified: bool = False
    verified_provenance: str = "no_pass"


class PipelineActionRequest(BaseModel):
    """Director action body for ``POST /pipeline/{version_id}/action``.

    ``action`` and ``payload`` shape are validated by the orchestrator
    (F-007 §5.2); invalid values surface as HTTP 4xx via the router.
    """

    action: str
    payload: Optional[dict[str, Any]] = None


class PipelineRelayRequest(BaseModel):
    """Body for ``POST /pipeline/{version_id}/relay`` (CR-V2-015) — a Manažér message typed in the
    read-only AI Agent tab.

    SPIKE-IO Model B: the message is RELAYED by the engine as the next ``--resume`` turn (the engine is the
    sole writer to the warm ``claude`` session) — it is NEVER keystroked into the PTY. When a turn is in
    flight the message is enqueued behind it; when settled it dispatches immediately as an ``ask``/``answer``.
    """

    text: str = Field(..., min_length=1, description="The Manažér's message to relay to the AI Agent.")


class PipelineRelayResponse(BaseModel):
    """Result of a relay: whether the message was ENQUEUED behind an in-flight turn (``deferred``) or
    dispatched immediately, plus the current board snapshot."""

    deferred: bool
    board: PipelineBoardRead


class FastFixStartRequest(BaseModel):
    """Body for ``POST /pipeline/fast-fix`` (F-009, CR-NS-094) — the "Rýchla oprava" entry.

    One prompt: ``project_id`` + the Director ``directive`` (the whole fix brief). The backend
    auto-creates the next PATCH version and starts a ``fast_fix`` pipeline carrying the directive.
    """

    project_id: UUID
    directive: str = Field(..., min_length=1, description="The Director's fast-fix directive (the task brief).")


class FastFixStartResponse(BaseModel):
    """Result of starting a Fast-Fix: the new PATCH ``version_id`` + the initial board snapshot."""

    version_id: UUID
    board: PipelineBoardRead
